xc7/libraries/parse_pdf_modules.py

```diff
@@ -0,0 +1,447 @@
+    # initialise the pdfminer interface --
+    # we use a custom "render device" to receive the text objects in the PDF for further processing
+    resman = PDFResourceManager()
+    doc = PDFDocument(PDFParser(open(infile, 'rb')))
+    laparams = LAParams(
+        line_margin=0.1, char_margin=0.7
+    )  # parameters optimised to prevent incorrectly joining together words
+
+    device = PDFTableParser(resman, laparams, stop_at='VHDL')
+    interpreter = PDFPageInterpreter(resman, device)
+
+    # parse the PDF table of contents to figure out what modules exist and which pages to process
+    port_pages = parse_module_pages(doc, 'Port Desc')
+    if not len(modules):
+        modules = port_pages.keys()  # default to processing ALL modules
+    attrib_list = parse_module_pages(
+        doc, 'Available Attrib')  # NB: not all modules have attributes
+
+    # parse the specifications and generate an XML tree
+    E = objectify.ElementMaker(annotate=False)
+    root = E.xml(source=infile, processed=datetime.now().isoformat())
+
+    # run through the modules
+    for module in modules:
+        sys.stderr.write('Processing %s...\n' % module)
+        node = E.module(name=module)
+        # process the ports of this module
+        for pg, top, bottom in port_pages[module]:
+            device.reset(top, bottom)
+            interpreter.process_page(pg)
+            for P in process_ports(device):
+                node.append(
+                    E.port(
+                        P['Function'],
+                        name=P['Port'],
+                        type=P['Type'],
+                        width=str(P['Width'])))
+            if device.done: break
+        # process the attributes of this module
+        for pg, top, bottom in attrib_list.get(module, []):
+            device.reset(top, bottom)
+            interpreter.process_page(pg)
+            for A in process_attributes(device):
+                node.append(
+                    E.attribute(
+                        A['Description'],
+                        name=A['Attribute'],
+                        type=A['Type'],
+                        default=A['Default'].replace('"', ''),
+                        values=A['Allowed'].replace('"', '')))
+            if device.done: break
+        # add it to the root object
+        root.append(node)
+    return root
+
+
+if __name__ == '__main__':
+    import argparse
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--input',
+        '-i',
+        nargs='?',
+        default='ug953-vivado-7series-libraries.pdf')
+    parser.add_argument(
+        '--output',
+        '-o',
+        nargs='?',
+        type=argparse.FileType('wb'),
+        default=sys.stdout)
+    parser.add_argument('--modules', '-m', nargs='*')
+    args = parser.parse_args()
+
+    xml = process_specs(args.input, args.modules)
+    args.output.write(etree.tostring(xml, pretty_print=True))
```

Comment:
I received an error when trying stdout. Apparently etree.tostring() returns bytes and to output to stdout you need to decode.
```suggestion
    args.output.write(etree.tostring(xml, pretty_print=True).decode('ascii'))
```